### PR TITLE
Use a specific version of an external git source dependency

### PR DESCRIPTION
**Pull request by eljobe**
This change allows repositories to be cloned at a specific tag, or
branch if the name matches the exact name of the version of the
dependency being resolved.

Fixes gradle/gradle-native#185

### CODE REVIEW

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/vcs/VcsDependencyResolver.java [hunk 1]
```diff
@@ -71,13 +71,19 @@ public void resolve(DependencyMetadata dependency, BuildableComponentIdResolveRe
...
             vcsMappingsInternal.getVcsMappingRule().execute(vcsMappingInternal);
 
             // TODO: Need failure handling, e.g., cannot clone repository
             if (vcsMappingInternal.hasRepository()) {
                 VersionControlSpec spec = vcsMappingInternal.getRepository();
                 VersionControlSystem versionControlSystem = versionControlSystemFactory.create(spec);
-                VersionRef selectedVersion = selectVersionFromRepository(spec, versionControlSystem);
+                VersionRef selectedVersion = selectVersionFromRepository(spec, versionControlSystem, depSelector.getVersion());
+                if (selectedVersion == null) {
+                    result.failed(new ModuleVersionResolveException(depSelector, spec.getDisplayName() + " does not contain a version matching " + depSelector.getVersion()));
```
Comment (big-guy) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/vcs/VcsDependencyResolver.java]: This should contain some context about what we were trying to resolve.  It currently looks like:
> Git Repository at file:/Users/sterling/gits/gradle/build/tmp/test%20files/GitVcsIntegrationTest/unknown-test/2tn5u/dep/ does not contain a version matching 1.4.0

It should look like the message from other `ModuleVersionResolveException` failures:
> "Could not resolve $depSelector. Git Repository at file:/Users/sterling/gits/gradle/build/tmp/test%20files/GitVcsIntegrationTest/unknown-test/2tn5u/dep/ does not contain a version matching '1.4.0'

File: subprojects/version-control/src/main/java/org/gradle/vcs/git/internal/GitVersionControlSystem.java [hunk 2]
```diff
@@ -71,7 +71,13 @@ public File populate(File versionDir, VersionRef ref, VersionControlSpec spec) {
         }
         Set<VersionRef> versions = Sets.newHashSet();
         for (Ref ref : refs) {
-            versions.add(GitVersionRef.from(ref));
+            GitVersionRef gitRef = GitVersionRef.from(ref);
+            versions.add(gitRef);
+            // The HEAD reference in a Git Repository is a logical choice if the user is looking
+            // for the 'latest.integration' version of a dependency.
+            if (gitRef.getVersion().equals("HEAD")) {
+                versions.add(GitVersionRef.from("latest.integration", gitRef.getCanonicalId()));
```
Comment (big-guy) [inline subprojects/version-control/src/main/java/org/gradle/vcs/git/internal/GitVersionControlSystem.java]: This breaks `org.gradle.vcs.git.internal.GitVersionControlSystemSpec#can get versions`

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/vcs/internal/GitVcsIntegrationTest.groovy [hunk 3]
```diff
@@ -106,7 +106,65 @@ class GitVcsIntegrationTest extends AbstractVcsIntegrationTest {
...
+        repo.commit('Changed Dep to an interface', GFileUtils.listFiles(file('dep'), null, true))
+
+        buildFile.text = buildFile.text.replace('latest.integration', '1.4.0')
+
+        when:
+        fails('assemble')
+
+        then:
+        failureCauseContains("does not contain a version matching 1.4.0")
     }
```
[the remark this answers is not in the document]
Comment (eljobe) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/vcs/internal/GitVcsIntegrationTest.groovy]: I added the test to demonstrate that revisiting a repository happens between builds. Also added a test for what happens when we have duplicate dependencies. I'm doubtful we're doing what we ultimately want to do. But, there is a test for it now. I'm not very familiar with the "force component selection" rules, so I'm sure we'll have to look more into getting that working.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/vcs/VcsDependencyResolver.java (hunk 1 above)
Comment (eljobe) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/vcs/VcsDependencyResolver.java]: Done

File: subprojects/version-control/src/main/java/org/gradle/vcs/git/internal/GitVersionControlSystem.java (hunk 2 above)
Comment (eljobe) [reply subprojects/version-control/src/main/java/org/gradle/vcs/git/internal/GitVersionControlSystem.java]: Fixed

**Pull request merged**